tests/suites/test_suite_psa_crypto_init.function

```diff
@@ -282,3 +282,73 @@ exit:
     mbedtls_psa_crypto_free( );
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PSA_CRYPTO_C */
+void key_policy_init()
+{
+    psa_key_policy_t func = psa_key_policy_init();
+    psa_key_policy_t init = PSA_KEY_POLICY_INIT;
+    psa_key_policy_t zero;
+
+    memset(&zero, 0, sizeof(zero));
+
+    TEST_EQUAL(memcmp(&func, &zero, sizeof(zero)), 0);
+    TEST_EQUAL(memcmp(&init, &zero, sizeof(zero)), 0);
+}
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PSA_CRYPTO_C */
+void hash_operation_init()
+{
+    psa_hash_operation_t func = psa_hash_operation_init();
+    psa_hash_operation_t init = PSA_HASH_OPERATION_INIT;
+    psa_hash_operation_t zero;
+
+    memset(&zero, 0, sizeof(zero));
+
+    TEST_EQUAL(memcmp(&func, &zero, sizeof(zero)), 0);
```

Comment:
I know what the standard says, but I'd love to see a platform where this actually breaks. When it does break, I really want to know about it because I'm sure we'll have other subtle portability issues on that platform. This test fails loudly when such a platform shows up.

Our current implementation depends (perhaps beyond what the C standard promises) on being able to initialize an entire struct that contains a union without clang initializer warnings by only initializing the smaller, simpler type (the `dummy` variable in the union) and not having union padding go uninitialized. It'd be good to have a test to catch if this dependency breaks due to new platforms or toolchains or versions thereof.

Reply:
I agree we should instead test how the structure acts with each initializer, not how its represented, as we technically don't care how it's represented (and couldn't control for its representation portably even if we did). We can do that in another PR.

Regarding your example, I don't believe compilers are allowed to avoid initializing the other fields of the struct or union, but they probably could leave any padding uninitialized. It would be interesting to observe such a case, or to see that the standard requires padding to also have static initialization the same as any other not-specified initializer.

Separate from the fact that we are testing the representation of all values logical zero as all-bits-zero, what our current initializer implementation depends on today is that:

Given:
```C
struct s {
  union {
    unsigned dummy;
    struct complicated {
        unsigned x;
        unsigned y;
#ifdef SOME_CONFIG_OPTION
        unsigned z;
#endif
    };
};
```
and
```C
struct s s = {0, {0}};
```
that
```C
s.y == 0
```

We depend on this because` {0, {0}}` is easier to have as "the one initializer that works everywhere even with Clang 5 and `-Wmissing-field-initializers`" instead of multiple different initializers for each of the configuration options. Technically, setting `struct s s = {0};` should also work just as well per the standard...

I'll update the test to check that the values are 0 instead of all-bits-zero, even when statically initialized. This is the minimum we depend on. We don't technically require that the entire object is all-bits-zero, just that its values are.

Reply:
If you're thinking of testing what the library code relies on, our code doesn't depend on the union part of the operation struct being initialized at all. All we need during initialization is to initialize the `alg` fields and some other algorithm-independent fields.

But you can't test this from the unit tests, because they need to work with different implementations of `crypto_struct.h` that have a different layout. When the library is integrated as a PSA service, each operation struct is `{psa_handle_t handle;}`.

The library never depends on type punning in unions, as far as I know. Doing so is dangerous because C89 didn't allow it and compilers in C89 mode sometimes decide that if you're writing to one union member and reading from another union member, these two can't possibly be aliased. What the library relies on is that if a data structure is initialized to all-bits-zero (by `calloc`) then all of its integer members have the logical value 0 (which is the case because we also forbid padding bits in integers) and all of its pointer members are null (we require that all-bits-zero is null). These implementation properties (no padding bits, all-bits-zero is null) are tested in `selftest.c`.

Reply:
Hmm, I don't think it's actually worth the time to update the test to check for all values logical zero for each of the initializers, given that we'll move to testing that the structs do what we want (vs are what we want). What you've requested is a weakening of a test, because we are testing for something more strict than what the standard allows compilers to do. But, if what we are testing for is true, then our weaker actual dependency will be met as well.

Reply:
> different implementations of crypto_struct.h that have a different layout

Our unit tests are specific to our implementation, not meant to work with other implementations. That's what the ACK is for. But, given that we'd like to use our tests with the service implementation, yes, we can't check that all values are logical zero because that's heavily dependent on the structs defined by `crypto_struct.h`. That's yet another reason to prefer testing based on what freshly initialized structs do versus how they are represented. Let's make this change to testing what the structs do in a separate PR.

Reply:
> I'd love to see a platform where this actually breaks.

MSVC then proceeded to immediately break. PR to fix this coming soon.